pipeline_dp/pipeline_operations.py

```diff
@@ -147,29 +150,29 @@ def count_per_element(self, rdd, stage_name: str = None):
 class LocalPipelineOperations(PipelineOperations):
     """Local Pipeline adapter."""
 
-    def map(self, col, fn, stage_name: str = None):
+    def map(self, col, fn, stage_name: Optional[str] = None):
         return map(fn, col)
 
     def map_tuple(self, col, fn, stage_name: str = None):
         return (fn(k, v) for k, v in col)
 
-    def map_values(self, col, fn, stage_name: str):
-        pass
+    def map_values(self, col, fn, stage_name: Optional[str] = None):
+        return ((k, fn(v)) for k, v in col)
 
     def group_by_key(self, col, stage_name: str):
         pass
 
-    def filter(self, col, fn, stage_name: str):
-        pass
+    def filter(self, col, fn, stage_name: Optional[str] = None):
+        return filter(fn, col)
 
     def keys(self, col, stage_name: str):
         pass
 
-    def values(self, col, stage_name: str):
-        pass
+    def values(self, col, stage_name: Optional[str] = None):
+        return (v for k, v in col)
 
     def sample_fixed_per_key(self, col, n: int, stage_name: str):
         pass
 
-    def count_per_element(self, col, stage_name: str):
-        pass
+    def count_per_element(self, col, stage_name: Optional[str] = None):
+        yield from collections.Counter(col).items()
```

Comment:
I like idea with Counter!

One question, in this line, is 'col' accessed the first time only when iterations over count_per_element() started?

What do you think about adding the 'lazyness' test:
Let 'exception_generator' is a generator which raise an exception on the first access to it

and then
def test_lazyness():
  self.ops.map(exception_generator) # should be no exception
  self.ops.map_tuple(exception_generator)
   ....

on implementing a new method it would be easy to add to the test and to ensure that the method is lazy

Reply:
Cool idea for testing!
I added a test like this and yes, `count_per_element` works as expected.